File: 2021/day_12.py
from typing import Dict, List


def extract_paths(lines: List[str]) -> Dict[str, List[str]]:
    caves = {}
    for line in lines:
        start, end = line.split("-")
        if start not in caves:
            caves[start] = [end]
        else:
            caves[start].append(end)
        if end not in caves:
            caves[end] = [start]
        else:
            caves[end].append(start)
    return caves


def visit(
        caves: Dict[str, List[str]],
        cave: str,
        visited: List[str],
        used_extra_cave: bool
) -> int:
    if cave in visited:
        if used_extra_cave or cave == "start":
            return 0
        else:
            used_extra_cave = True
    if cave == "end":
        return 1
    paths = 0
    if cave.islower():
        visited.append(cave)
    for dest in caves.get(cave, []):
        paths += visit(caves, dest, visited, used_extra_cave)
    if cave.islower():
        visited.remove(cave)
    return paths


def part1(data: str):
    lines = data.splitlines()
    caves = extract_paths(lines)
    return visit(caves, "start", [], True)


def part2(data: str):
    lines = data.splitlines()
    caves = extract_paths(lines)
    return visit(caves, "start", [], False)


if __name__ == "__main__":
    test = True
    test = False
    test_input = """fs-end
he-DX
fs-he
start-DX
pj-DX
end-zg
zg-sl
zg-pj
pj-he
RW-he
fs-DX
pj-RW
zg-RW
start-pj
he-WI
zg-he
pj-fs
start-RW"""
    if test:
        puzzle_input = test_input
    else:
        with open("day_12_input.txt", "r") as input_file:
            puzzle_input = input_file.read().strip()
    print(part1(puzzle_input))
    print(part2(puzzle_input))

File: 2017/day_12.rs
//! https://adventofcode.com/2017/day/12
//! https://adventofcode.com/2017/day/12/input

use std::{
    collections::{HashSet, VecDeque},
    fs::read_to_string,
    time::Instant,
};

type Graph = Vec<HashSet<usize>>;

fn parse(input: &str) -> Graph {
    input
        .lines()
        .map(|line| {
            let mut parts = line.split(" <-> ");
            let pipes = parts.nth(1).unwrap();
            pipes.split(", ").map(|p| p.parse().unwrap()).collect()
        })
        .collect()
}

fn bfs(graph: &Graph, visited: &mut Vec<bool>) -> usize {
    let mut count = 0;
    let mut queue = VecDeque::from([(0..visited.len()).find(|&i| !visited[i]).unwrap()]);
    while let Some(program) = queue.pop_front() {
        visited[program] = true;
        count += 1;
        for &pipe in &graph[program] {
            if !visited[pipe] {
                queue.push_back(pipe);
            }
        }
    }
    count
}

pub mod part1 {
    use super::{bfs, parse};

    pub fn solve(input: &str) -> usize {
        let graph = parse(input);
        let mut visited = vec![false; graph.len()];
        bfs(&graph, &mut visited)
    }
}

pub mod part2 {
    use super::{bfs, parse};

    pub fn solve(input: &str) -> usize {
        let graph = parse(input);
        let mut count = 0;
        let mut groups = 0;
        let mut visited = vec![false; graph.len()];
        while count < graph.len() {
            groups += 1;
            count += bfs(&graph, &mut visited)
        }
        groups
    }
}

pub fn main(test: bool) {
    let test_input = "".to_owned();
    let puzzle_input = if test {
        test_input
    } else {
        read_to_string("inputs/day_12_input.txt").unwrap()
    };
    let start = Instant::now();
    println!("{}", part1::solve(&puzzle_input));
    println!("Run in {:?}", start.elapsed());
    let start = Instant::now();
    println!("{}", part2::solve(&puzzle_input));
    println!("Run in {:?}", start.elapsed());
}
